Week2/homework.js

```diff
@@ -0,0 +1,164 @@
+const mysql = require('mysql');
+const { promisify } = require('util');
+const inquirer = require('inquirer');
+
+// Using virtual ubuntu server on local running mysql instance => 192.168.153.132 IP of VM
+const conn = mysql.createConnection('mysql://hyfuser:hyfpassword@192.168.153.132/new_world');
+
+conn.connect();
+
+const runQuery = promisify(conn.query).bind(conn);
+
+//* What is the capital of country X ? (Accept X from user)
+const getCapitalName = async () => {
+  const question = [
+    {
+      type: 'input',
+      name: 'countryName',
+      message: 'Country name?',
+    },
+  ];
+  const { countryName } = await inquirer.prompt(question);
+  const result = await runQuery(
+    `
+    SELECT 
+      city.name AS capital
+    FROM
+      country,
+      city
+    WHERE
+      country.capital = city.id
+        AND country.name = ?`,
+    countryName,
+  );
+  if (!result.length) throw new Error('Wrong country name');
+  console.log(`The capital city of ${countryName} is ${result[0].capital}`);
+};
+
+//* List all the languages spoken in the region Y (Accept Y from user)
+const getLangsOfRegion = async () => {
+  const question = [
+    {
+      type: 'input',
+      name: 'regionName',
+      message: 'Region name?',
+    },
+  ];
+  const { regionName } = await inquirer.prompt(question);
+  const result = await runQuery(
+    `
+    SELECT DISTINCT
+      countrylanguage.language
+    FROM
+      country,
+      countrylanguage
+    WHERE
+      countrylanguage.countrycode = country.code
+        AND country.region = ?
+    ORDER BY language`,
+    regionName,
+  );
+  if (!result.length) throw new Error('Wrong region name');
+  console.log(`The languages spoken in ${regionName} : ${result.map(c => c.language).toString()}`);
+};
+
+//*Find the number of cities in which language Z is spoken (Accept Z from user)
+const getNumberOfSpokenCities = async () => {
+  const question = [
+    {
+      type: 'input',
+      name: 'language',
+      message: 'Language?',
+    },
+  ];
+  const { language } = await inquirer.prompt(question);
+  const result = await runQuery(
+    `
+    SELECT 
+      countrylanguage.language, COUNT(city.name) numberOfCities
+    FROM
+      country,
+      city,
+      countrylanguage
+    WHERE
+      country.code = city.countrycode
+        AND country.code = countrylanguage.countrycode
+        AND countrylanguage.language = ?
+    GROUP BY language;
+    `,
+    language,
```

Comment:
It's working but you can make it a lot easier, do you need to join on the three tables ?

Reply:
I tried with subquery but still using three tables. and become more complex. Am i missing something? As for the references between three tables i couldn't see any solution without using all of them.

```
SELECT 
    language, SUM(result.cnt)
FROM
    countrylanguage,
    (SELECT 
        country.code countryCode, COUNT(city.name) cnt
    FROM
        country, city
    WHERE
        country.Code = city.CountryCode
    GROUP BY countryCode) result
WHERE
    result.countrycode = countrylanguage.CountryCode
GROUP BY language
ORDER BY 2 DESC;
```

Reply:
What we want here is the number of city speaking a language right , So it doesn't involved Country . You can only use city and countrylanguage, joining on countrycode (countrycode is in city AND countrylanguage) where the language is <?>

Reply:
You are totally right. Thanks for the tip. I missed the point that city and countrylanguage tables have both countrycode column. This query is the easier one with two tables: 

```
SELECT 
      countrylanguage.language, COUNT(city.name) numberOfCities
    FROM
      city,
      countrylanguage
    WHERE city.countrycode = countrylanguage.countrycode
        AND countrylanguage.language = 'English'
    GROUP BY language;
```

Reply:
Almost, you can do even simpler, no need of the group by here :)

Reply:
Making a group by over a column which is in the where clause and also selecting it... That sounds meaningless 😀
Hope this is all done now: 
```
SELECT 
      COUNT(city.name) numberOfCities
    FROM
      city,
      countrylanguage
    WHERE city.countrycode = countrylanguage.countrycode
        AND countrylanguage.language = 'English';
```